cip/1.accepted/CIP2021-07-07-Grouping-keys-and-aggregation-expressions.adoc

```diff
@@ -0,0 +1,779 @@
+----
+
+Analogously, `WITH _p_~1~, _p_~2~, ..., _p~N~_ _ORDER-SKIP-LIMIT_` is effectively equivalent to
+
+[source, cypher, subs="quotes"]
+----
+WITH _PRE_PROJ_(_p_~1~), _PRE_PROJ_(_p_~2~), ..., _PRE_PROJ_(_p~N~_)
+WITH _AGGR_(_p_~1~), _AGGR_(_p_~2~), ..., _AGGR_(_p~N~_)
+WITH _POST_PROJ_(_p_~1~), _POST_PROJ_(_p_~2~), ..., _POST_PROJ_(_p~N~_) _ORDER-SKIP-LIMIT_
+----
+====
+
+Every https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=SortItem[<SortItem>] listed in the ORDER BY clause contains an expression.
+Since these expressions are effectively evaluate after all _POST_PROJ_(_p~i~_) expressions, a similar syntax restrictions applies to the <SortItem>s.
+
+However, <SortItem>s can refer to aliases introduced by _POST_PROJ_(_p~i~_).
+
+Given a list of <ProjectionItem>s _L_ = {_p_~1~, _p_~2~, ..., _p~N~_}, let _GROUPING_KEYS_AND_ALIASES_(_L_) be the set of
+
+* all expressions _ex_ for which there is a <ProjectionItem>s _p_ = `_ex_ AS _al_` in _L_ where _AGG_(_p_) is empty and
+* all aliases _al_ for which there is a <ProjectionItem>s _p_ = `_ex_ AS _al_` in _L_.
+
+For an expression _ex_ and a projection list _L_, let _AVAILABLE_TO_ORDER_(_ex_, _L_) hold
+
+* If _ex_ is either
+** A grouping key or an alias, i.e. _ex_ ∈ _GROUPING_KEYS_AND_ALIASES_(_L_),
+** A constant, or
+** A parameter,
+* or if _ex_ comprises of sub-expressions, it only comprises sub-expressions _subEx_ for which _AVAILABLE_TO_ORDER_(_subEx_, _L_) holds.
+
+[IMPORTANT]
+====
+For `WITH _L_ ORDER BY _SI_` and `RETURN _L_ ORDER BY _SI_` and every _ex_ contained in a <SortItem> in _SI_, _AVAILABLE_TO_ORDER_(_ex_, _L_) shall hold.
+====
+
+==== Omitted aliases
+
+This proposal considers all <ProjectionItem>s have user-given alias.
+Cypher allows to omit the aliases, particularly in the RETURN clause, though.
+However, the alias omission rules are based on the assumption that an implementation will infer a more or less reasonable alias if an alias is omitted by the user.
+Hence, it is safe for this proposal to assume that all <ProjectionItem>s have an alias.
+
+=== Examples
+
+==== Valid aggregations
+
+The following clauses exhibit valid aggregations under the grouping and aggregation semantics and the syntax restriction it includes.
+For each example we list why it is valid.
+
+. `RETURN 1 + count(*)`
+* The sub-expression `1` is a constant.
+
+. `RETURN 1, 1 + count(*)`
+* The sub-expression `1` is a constant.
+
+. `RETURN $x + count($x)`
+* The sub-expression `$x` is a parameter.
+
+. `RETURN count($x) + $x`
+* The sub-expression `$x` is a parameter.
+
+. `RETURN 1 + count($x) + $x * 7 + sum($x) + 'cake'`
+* The sub-expressions `1`, `2`, and `'cake'` are constants.
+* The sub-expression `$x` is a parameter.
+
+. `MATCH (a) RETURN a.x, 1 + count(a.x)`
+* The sub-expression `1` is a constant.
+
+. `MATCH (a) RETURN a.x, a.x + count(a.x)`
```

Comment:
As mentioned in another comment as well. Didn't we end up with the rules:

```
In an expression that contains aggregation, the leaves of expressions must be one of:
An aggregation
A literal
A Parameter
A variable - ONLY IF that variable is also a projection expression on its own (e.g. the n in RETURN n AS myNode, n.value + count(*))
```

In this expression `a` is the leaf and it is _not_  a projection expression on its own

Reply:
Well, initially I thought so too.

However, attached to one CLG card is implementation report pdf attached to this card
https://trello.com/c/56tgXSFP/326-grouping-key-confusion-when-combining-aggregating-and-non-aggregating-values
(It isn't a google doc unfortunately for some reason, so it is unclear if there is newer version of it.)

This document lists exactly this query `MATCH (a) RETURN a.x, a.x + count(a.x)` as number 7 in the section called "Ok (Not deprecated)".

Going pure leaves would certainly be cleaner. But it would als deprecate even more query. Not everyone would be happy about that as you can imaging.

Reply:
I guess we need to bring this up, to make it more clear how it should be. 

(On the 6th of August I wrote, on the same card as you just linked, a couple of examples where `MATCH (a) RETURN a.x, a.x + count(a.x)` was one of the non-allowed queries)